app/services/data_decoder.py

```diff
@@ -0,0 +1,393 @@
+        data_decoder_service = DataDecoderService()
+        await data_decoder_service.init(session)
+    return data_decoder_service
+
+
+class DataDecoderService:
+    EXEC_TRANSACTION_SELECTOR = HexBytes("0x6a761202")
+
+    dummy_w3 = Web3()
+    session: AsyncSession | None
+
+    async def init(self, session: AsyncSession) -> None:
+        """
+        Initialize the data decoder service, loading the ABIs from the database and storing the 4byte selectors
+        in memory
+        """
+        logger.info("%s: Loading contract ABIs for decoding", self.__class__.__name__)
+        self.fn_selectors_with_abis: dict[bytes, ABIFunction] = (
+            await self._generate_selectors_with_abis_from_abis(
+                await self.get_supported_abis(session)
+            )
+        )
+        logger.info(
+            "%s: Contract ABIs for decoding were loaded", self.__class__.__name__
+        )
+        self.multisend_abis: list[ABI] = [m async for m in self.get_multisend_abis()]
+        self.multisend_fn_selectors_with_abis: dict[bytes, ABIFunction] = (
+            await self._generate_selectors_with_abis_from_abis(
+                self.get_multisend_abis()
+            )
+        )
+
+    def _generate_selectors_with_abis_from_abi(
+        self, abi: ABI
+    ) -> dict[bytes, ABIFunction]:
+        """
+        :param abi: ABI
+        :return: Dictionary with function selector as bytes and the ContractFunction
+        """
+        return {
+            function_abi_to_4byte_selector(cast(dict[str, Any], fn_abi)): fn_abi
+            for fn_abi in abi
+            if fn_abi["type"] == "function"
+        }
+
+    async def _generate_selectors_with_abis_from_abis(
+        self, abis: AsyncIterator[ABI]
+    ) -> dict[bytes, ABIFunction]:
+        """
+        :param abis: Contract ABIs. Last ABIs on the Sequence have preference if there's a collision on the
+        selector
+        :return: Dictionary with function selector as bytes and the function abi
+        """
+        return {
+            fn_selector: fn_abi
+            async for supported_abi in abis
+            for fn_selector, fn_abi in self._generate_selectors_with_abis_from_abi(
+                supported_abi
+            ).items()
+        }
+
+    async def get_supported_abis(self, session: AsyncSession) -> AsyncIterator[ABI]:
+        """
+        :return: Every ABI in the database
+        """
+        return Abi.get_abis_sorted_by_relevance(session)
+
+    async def get_multisend_abis(self) -> AsyncIterator[ABI]:
+        yield get_multi_send_contract(self.dummy_w3).abi
+
+    @alru_cache(maxsize=2048)
+    async def get_contract_abi(self, address: Address) -> ABI | None:
+        """
+        Retrieves the ABI for the contract at the given address.
+
+        :param address: Contract address
+        :return: List of ABI data if found, `None` otherwise.
+        """
+        async with AsyncSession(get_engine(), expire_on_commit=False) as session:
```

Comment:
Why this function is not using getting the session from database functions? you can use the decorator too.